### ISSUE TITLE
IWYU uses Xcode headers

### ISSUE THREAD

**Issue by Timmmm** [include-what-you-use/include-what-you-use#438]
The tool seems to be giving me weird results that I would have thought were resolved by the mapping system. For example on this source:

```
#include <string>
#include <iostream>

int main()
{
	std::string hello = "Hello world";
	std::cout << hello << std::endl;
	return 0;
}
```

It tells me:

```
In file included from main.cpp:1:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/string:439:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:624:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/initializer_list:47:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/cstddef:38:
In file included from /usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1/stddef.h:51:
/usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1/__nullptr:24:29: error: variable has incomplete type 'struct _LIBCPP_TEMPLATE_VIS'
struct _LIBCPP_TEMPLATE_VIS nullptr_t
                            ^
/usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1/__nullptr:24:8: note: forward declaration of 'std::__1::_LIBCPP_TEMPLATE_VIS'
struct _LIBCPP_TEMPLATE_VIS nullptr_t
       ^
/usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1/__nullptr:24:38: error: expected ';' after top level declarator
struct _LIBCPP_TEMPLATE_VIS nullptr_t
                                     ^
/usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1/__nullptr:47:48: error: unknown type name 'nullptr_t'
inline _LIBCPP_ALWAYS_INLINE _LIBCPP_CONSTEXPR nullptr_t __get_nullptr_t() {return nullptr_t(0);}
                                               ^
In file included from main.cpp:2:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/iostream:40:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/istream:163:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/ostream:140:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/locale:192:
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdlib:167:44: error: declaration conflicts with target of using declaration already in scope
inline _LIBCPP_INLINE_VISIBILITY long      abs(     long __x) _NOEXCEPT {return  labs(__x);}
                                           ^
/usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1/stdlib.h:115:44: note: target of using declaration
inline _LIBCPP_INLINE_VISIBILITY long      abs(     long __x) _NOEXCEPT {return  labs(__x);}
                                           ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdlib:135:9: note: using declaration
using ::abs;
        ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdlib:169:44: error: declaration conflicts with target of using declaration already in scope
inline _LIBCPP_INLINE_VISIBILITY long long abs(long long __x) _NOEXCEPT {return llabs(__x);}
                                           ^
/usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1/stdlib.h:117:44: note: target of using declaration
inline _LIBCPP_INLINE_VISIBILITY long long abs(long long __x) _NOEXCEPT {return llabs(__x);}
                                           ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdlib:135:9: note: using declaration
using ::abs;
        ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdlib:172:42: error: declaration conflicts with target of using declaration already in scope
inline _LIBCPP_INLINE_VISIBILITY  ldiv_t div(     long __x,      long __y) _NOEXCEPT {return  ldiv(__x, __y);}
                                         ^
/usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1/stdlib.h:120:42: note: target of using declaration
inline _LIBCPP_INLINE_VISIBILITY  ldiv_t div(     long __x,      long __y) _NOEXCEPT {return  ldiv(__x, __y);}
                                         ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdlib:143:9: note: using declaration
using ::div;
        ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdlib:174:42: error: declaration conflicts with target of using declaration already in scope
inline _LIBCPP_INLINE_VISIBILITY lldiv_t div(long long __x, long long __y) _NOEXCEPT {return lldiv(__x, __y);}
                                         ^
/usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1/stdlib.h:122:42: note: target of using declaration
inline _LIBCPP_INLINE_VISIBILITY lldiv_t div(long long __x, long long __y) _NOEXCEPT {return lldiv(__x, __y);}
                                         ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdlib:143:9: note: using declaration
using ::div;
        ^

main.cpp should add these lines:

main.cpp should remove these lines:
- #include <string>  // lines 1-1

The full include-list for main.cpp:
#include <iostream>  // for operator<<, string, basic_ostream, cout, endl
---
```

I think this is caused by it jumping to the `/Library/Developer/CommandLineTools/usr/include/c++/v1/` headers. I even tried changing my `$PATH` to put `/usr/local/opt/llvm/bin` first. Here's how I install IWYU (on OSX):

1. `brew install llvm`
2. `git clone https://github.com/include-what-you-use/include-what-you-use.git`
3. `cd include-what-you-use`
4. `git checkout clang_4.0`
5. `cd ..`
6. `mkdir build`
7. `cd build`
8. `cmake -DIWYU_LLVM_ROOT_PATH=/usr/local/opt/llvm ../include-what-you-use`
9. `make -j4`
10. `cp include-what-you-use /usr/local/opt/llvm/bin/`

And then I run it directly on `main.cpp`.

The thing is I have no idea how it is getting to `/Library/Developer/CommandLineTools`. There are no environment variables that point to it, and nothing in `/usr/include/c++` seems to symlink to it. I *have* selected it with `xcode-select`, but why would IWYU care about that? Is it hard-coded to use `/usr/bin/clang++` or something?

**Comment by vsapsai** [comment]
Can you please run include-what-you-use on your main.cpp file with `-v` flag? It should print header search paths and we should be able to diagnose why `../include/c++/v1` (`/usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1` in your case) isn't selected before `/Library/Developer/CommandLineTools`.

We add libc++ headers bundled with Xcode.app to header search paths because otherwise clang (and IWYU) search for stdlib in `../include/c++/v1` and `/usr/include/c++/v1`. And developer tools provided by Apple don't install stdlib in those locations.

Thanks for explaining how you installed IWYU, it will help with troubleshooting the issue.

**Comment by Timmmm** [comment]
Ok so firstly the full output with `-v`.

```
$ /usr/local/opt/llvm/bin/include-what-you-use -v main.cpp
clang version 4.0.0 (tags/RELEASE_400/final)
Target: x86_64-apple-darwin16.6.0
Thread model: posix
InstalledDir: /usr/local/Cellar/llvm/4.0.0_1/bin
clang invocation:
 "/usr/local/Cellar/llvm/4.0.0_1/bin/include-what-you-use" "-cc1" "-triple" "x86_64-apple-macosx10.12.0" "-Wdeprecated-objc-isa-usage" "-Werror=deprecated-objc-isa-usage" "-fsyntax-only" "-disable-free" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "main.cpp" "-mrelocation-model" "pic" "-pic-level" "2" "-mthread-model" "posix" "-mdisable-fp-elim" "-masm-verbose" "-munwind-tables" "-target-cpu" "penryn" "-target-linker-version" "278.4" "-v" "-dwarf-column-info" "-debugger-tuning=lldb" "-resource-dir" "/usr/local/Cellar/llvm/4.0.0_1/bin/../lib/clang/4.0.0" "-stdlib=libc++" "-fdeprecated-macro" "-fdebug-compilation-dir" "/Users/me/local/build" "-ferror-limit" "19" "-fmessage-length" "160" "-stack-protector" "1" "-fblocks" "-fobjc-runtime=macosx-10.12.0" "-fencode-extended-block-signature" "-fcxx-exceptions" "-fexceptions" "-fmax-type-align=16" "-fdiagnostics-show-option" "-x" "c++" "main.cpp"

clang -cc1 version 4.0.0 based upon LLVM 4.0.0 default target x86_64-apple-darwin16.6.0
ignoring nonexistent directory "/usr/include/c++/v1"
#include "..." search starts here:
#include <...> search starts here:
 /Library/Developer/CommandLineTools/usr/include/c++/v1
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1
 /usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1
 /usr/local/include
 /usr/local/Cellar/llvm/4.0.0_1/bin/../lib/clang/4.0.0/include
 /usr/include
 /System/Library/Frameworks (framework directory)
 /Library/Frameworks (framework directory)
End of search list.
In file included from main.cpp:1:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/string:439:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/algorithm:624:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/initializer_list:47:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/cstddef:38:
In file included from /usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1/stddef.h:51:
/usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1/__nullptr:24:29: error: variable has incomplete type 'struct _LIBCPP_TEMPLATE_VIS'
struct _LIBCPP_TEMPLATE_VIS nullptr_t
                            ^
/usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1/__nullptr:24:8: note: forward declaration of 'std::__1::_LIBCPP_TEMPLATE_VIS'
struct _LIBCPP_TEMPLATE_VIS nullptr_t
       ^
/usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1/__nullptr:24:38: error: expected ';' after top level declarator
struct _LIBCPP_TEMPLATE_VIS nullptr_t
                                     ^
/usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1/__nullptr:47:48: error: unknown type name 'nullptr_t'
inline _LIBCPP_ALWAYS_INLINE _LIBCPP_CONSTEXPR nullptr_t __get_nullptr_t() {return nullptr_t(0);}
                                               ^
In file included from main.cpp:2:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/iostream:40:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/istream:163:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/ostream:140:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/locale:192:
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdlib:167:44: error: declaration conflicts with target of using declaration already in scope
inline _LIBCPP_INLINE_VISIBILITY long      abs(     long __x) _NOEXCEPT {return  labs(__x);}
                                           ^
/usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1/stdlib.h:115:44: note: target of using declaration
inline _LIBCPP_INLINE_VISIBILITY long      abs(     long __x) _NOEXCEPT {return  labs(__x);}
                                           ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdlib:135:9: note: using declaration
using ::abs;
        ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdlib:169:44: error: declaration conflicts with target of using declaration already in scope
inline _LIBCPP_INLINE_VISIBILITY long long abs(long long __x) _NOEXCEPT {return llabs(__x);}
                                           ^
/usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1/stdlib.h:117:44: note: target of using declaration
inline _LIBCPP_INLINE_VISIBILITY long long abs(long long __x) _NOEXCEPT {return llabs(__x);}
                                           ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdlib:135:9: note: using declaration
using ::abs;
        ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdlib:172:42: error: declaration conflicts with target of using declaration already in scope
inline _LIBCPP_INLINE_VISIBILITY  ldiv_t div(     long __x,      long __y) _NOEXCEPT {return  ldiv(__x, __y);}
                                         ^
/usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1/stdlib.h:120:42: note: target of using declaration
inline _LIBCPP_INLINE_VISIBILITY  ldiv_t div(     long __x,      long __y) _NOEXCEPT {return  ldiv(__x, __y);}
                                         ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdlib:143:9: note: using declaration
using ::div;
        ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdlib:174:42: error: declaration conflicts with target of using declaration already in scope
inline _LIBCPP_INLINE_VISIBILITY lldiv_t div(long long __x, long long __y) _NOEXCEPT {return lldiv(__x, __y);}
                                         ^
/usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1/stdlib.h:122:42: note: target of using declaration
inline _LIBCPP_INLINE_VISIBILITY lldiv_t div(long long __x, long long __y) _NOEXCEPT {return lldiv(__x, __y);}
                                         ^
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdlib:143:9: note: using declaration
using ::div;
        ^

main.cpp should add these lines:

main.cpp should remove these lines:
- #include <string>  // lines 1-1

The full include-list for main.cpp:
#include <iostream>  // for operator<<, string, basic_ostream, cout, endl
---
```

So that explains how it is finding the XCode headers. So are [these instructions](https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/InstructionsForUsers.md#how-to-install) totally wrong for OSX? Do I not need to do that at all?

I tried running it not from `/usr/local/opt/llvm/bin/`. It actually almost works, but can't find `stdarg.h` for some reason:

```
$ ./include-what-you-use -v main.cpp 
clang version 4.0.0 (tags/RELEASE_400/final)
Target: x86_64-apple-darwin16.6.0
Thread model: posix
InstalledDir: /Users/me/local/build
clang invocation:
 "/Users/me/local/build/include-what-you-use" "-cc1" "-triple" "x86_64-apple-macosx10.12.0" "-Wdeprecated-objc-isa-usage" "-Werror=deprecated-objc-isa-usage" "-fsyntax-only" "-disable-free" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "main.cpp" "-mrelocation-model" "pic" "-pic-level" "2" "-mthread-model" "posix" "-mdisable-fp-elim" "-masm-verbose" "-munwind-tables" "-target-cpu" "penryn" "-target-linker-version" "278.4" "-v" "-dwarf-column-info" "-debugger-tuning=lldb" "-resource-dir" "/Users/me/local/build/../lib/clang/4.0.0" "-stdlib=libc++" "-fdeprecated-macro" "-fdebug-compilation-dir" "/Users/me/local/build" "-ferror-limit" "19" "-fmessage-length" "160" "-stack-protector" "1" "-fblocks" "-fobjc-runtime=macosx-10.12.0" "-fencode-extended-block-signature" "-fcxx-exceptions" "-fexceptions" "-fmax-type-align=16" "-fdiagnostics-show-option" "-x" "c++" "main.cpp"

clang -cc1 version 4.0.0 based upon LLVM 4.0.0 default target x86_64-apple-darwin16.6.0
ignoring nonexistent directory "/Users/me/local/build/../include/c++/v1"
ignoring nonexistent directory "/usr/include/c++/v1"
ignoring nonexistent directory "/Users/me/local/build/../lib/clang/4.0.0/include"
#include "..." search starts here:
#include <...> search starts here:
 /Library/Developer/CommandLineTools/usr/include/c++/v1
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1
 /usr/local/include
 /usr/include
 /System/Library/Frameworks (framework directory)
 /Library/Frameworks (framework directory)
End of search list.
In file included from main.cpp:1:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/string:435:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/iosfwd:90:
/usr/include/wchar.h:89:10: fatal error: 'stdarg.h' file not found
#include <stdarg.h>
         ^~~~~~~~~~

main.cpp should add these lines:

main.cpp should remove these lines:
- #include <string>  // lines 1-1

The full include-list for main.cpp:
#include <iostream>  // for string
---

```

**Comment by vsapsai** [comment]
To solve your immediate problem with `'stdarg.h' file not found` you can copy `/usr/local/Cellar/llvm/4.0.0_1/bin/../lib/clang/4.0.0/include` next to include-what-you-use binary so it looks like

```
.
|- bin
|  `- include-what-you-use
`- lib
   `- clang
      `- 4.0.0
         `- include
            |- headers
            `- ...
```

This problem is described in Clang docs [here](https://clang.llvm.org/docs/FAQ.html#i-get-errors-about-some-headers-being-missing-stddef-h-stdarg-h).

As for the reason why you see the original problem with `error: variable has incomplete type 'struct _LIBCPP_TEMPLATE_VIS'` is that CLI and Xcode libc++ headers in header search paths come before libc++ headers for the installed llvm (`/usr/local/Cellar/llvm/4.0.0_1/bin/../include/c++/v1`). As the result IWYU encounters libc++ headers from 2 different versions - one shipped by Apple, another is public libc++. And turns out these versions are incompatible.

**Comment by jasonmp85** [comment]
Is there a fix for this? At minimum, the binary `include-what-you-use` download for darwin is entirely broken…

**Comment by kimgr** [comment]
Not that I'm aware of. I'm not sure I understand what the underlying problem is, even. Does the latest Xcode come with LLVM 4.0.1? Or does it just have its own naming convention for directories?

**Comment by jasonmp85** [comment]
See #484 for repro steps. I am not entirely sure of the root cause myself yet, but figure at minimum the provided binary builds should work (what is the intended use of the darwin binary you all vend? what are the requirements for using it?)

**Comment by kimgr** [comment]
To save people from downloading the entire LLVM+Clang repo and building from scratch. At one point that used to work pretty transparently (I think), but it's deteriorated with more frequent LLVM releases.

**Comment by kimgr** [comment]
I wonder if this might be fixed in 0.19 as of https://github.com/include-what-you-use/include-what-you-use/commit/24985e2ecf9365db41b5922fcb5ece076dea014f.

@Timmmm @jasonmp85 @ptomato If you guys have a way of testing 0.19, maybe you could ring back and let me know if it works out this issue for you? See the commit message for some details on customizing the behavior.

**Comment by ptomato** [comment]
0.19 does work out of the box on Homebrew, but it's hard to tell if that's because of this fix or because Homebrew now symlinks their LLVM headers to the expected location.

**Comment by kimgr** [comment]
@ptomato I've decided to take credit :smile:. Closing this as fixed in 0.19. Thanks for testing!
